### ISSUE TITLE
New features list/slide when Moonshine updates

### ISSUE THREAD

**Issue by rat-moonshine** [prominic/Moonshine-IDE#18]
It'd be nice if we can introduce a slide type of popup describing the new features, after Moonshine updates, during startup and only once.

**Comment by rat-moonshine** [comment]
I found this some time back while an app updated on my desktop and started for the first time (after get updated), the new/updated feature list (we probably need something simple like this) :

![whatsnew](https://user-images.githubusercontent.com/24370870/33867847-59737f66-df25-11e7-9437-84fd8249f5c4.png)

**Comment by piotrzarzycki21** [comment]
@rat-moonshine , @JoelProminic What do you think about displaying in that window content which is in GitHub CHANGELOG ? 

In that case:
- We will have to maximum change link to the CHANGELOG on each release
- We won't display that window if there will be some problem with parsing

**Comment by JoelProminic** [comment]
If we could automatically populate the updates from the GitHub CHANGELOG.MD, that would make it easier to ensure that we remember to add it for the release.

I think we might want to extract the CHANGELOG.md text for the latest version automatically as part of the build process rather than having Moonshine read the latest copy (or a linked copy) from GitHub.

Either way, we could also add a script to the build process that would test whether there were any parsing errors with the CHANGELOG.md file.  (I would need to find a way to make this available to the developers who do not have access to the Prominic Bamboo, though).

**Comment by piotrzarzycki21** [comment]
I'm not clear how are you going to parse it outside the Moonshine ? What kind of script do you mean ? After the parsing we will have some simpler for example XML file which contains last version information ?

**Comment by JoelProminic** [comment]
I was thinking of just some verification of the CHANGELOG.md format, if the parsing logic was simple.   If the parsing logic is complicated, then it would be a bit overkill to write the parsing logic in both a script and the Actionscript code.

Parsing and writing the CHANGELOG as XML could be a nice solution if the parsing logic is complex.

**Comment by piotrzarzycki21** [comment]
What do you think about following solution ?

1) Make script for bamboo which load CHANGELOG and parse it. It could pickup only the latest release part based on "Moonshine IDE [1.7.1]".
2) Create some result file XML/JSON or any other format: changelog.json - It contains only latest changes.
3) Script will save file in the elements/data/changelog.json 
4) Moonshine read it and easy transform in the ActionScript to the list similar on the screenshot.
5) Each update of Moonshine will have to reset somehow ability to showing that list.

If we go to that direction, what language should be used to provide such script for Bamboo ?

**Comment by rat-moonshine** [comment]
Why Bamboo involves into this? 

Please think that is only a cloud server where we build Moonshine, it should not be any dependency or to parse any business logic for the Moonshine. Moonshine needs to build is own strategy to parse whichever data from wherever.

**Comment by piotrzarzycki21** [comment]
Ok it can be in the Moonshine and in point 2/3 we don't have to save anything as file. Just display it.

**Comment by JoelProminic** [comment]
To review, here are some of the options to consider if we want to use CHANGELOG.md to populate the "New Features" window:
1. Bundling CHANGELOG.MD with the Moonshine installer.  This way, we won't need to worry about the CHANGELOG.md contents changing after the installer has been released (and introducing new parsing errors)
2. Verification of CHANGELOG.MD during Bamboo build.  This would be useful for catching potential parsing errors before we test the application.   If this window only shows up when a version is first installed, then it may be easy to forget to check the New Feature content before the release.
3. Parsing CHANGELOG.MD to an intermediate format.  This is only really necessary if we want to avoid duplicating the actual parsing logic with the verification logic above.

I think we should at least do #1 in the Ant/Bamboo build.  #2 could be added/improved later, since it is mostly just a testing convenience.

I like the idea of using CHANGELOG.md, but I would like to avoid the case where a new installation shows an obvious error for "New Features" (or crashes).

The external validator or parsing script could technically be written in any (reasonable) scripting language, but I'd prefer to keep it using on of our existing dependencies (Ant, Bash, Java) to avoid adding additional dependencies.

**Comment by piotrzarzycki21** [comment]
I agree with option **1** in ANT it would be easy bundle CHANGELOG.md as part of the installation.

Your thoughts on option **3** is not clear to me. Let's say that somehow in option **2** we will check whether format of CHANGELOG is correct. We will have to parse it later again to the format understandable for some Flex component which will display that text. Bold headers, give appropriate fonts to other things etc. We won't avoid at some point parsing duplication. 

If you meant in option *3* that in option **2** apart of checking we will produce some understandable format for Flex that's fine with me. 

Another thought is that, maybe some Flex component will just nicely display that text without any parsing.

**Comment by rat-moonshine** [comment]
I would like to revisit this issue. 

My suggestion is to use HTML component to display a file saved somewhere to the Moonshine website, or GitHub. This will also support the suggestion from @piotrzarzycki21 to not to parse content separately (at Flex side) but display as things come. 

However, this solution may not render CHANGELOG.md file's markup precisely; I think this should be in a consideration to not to parse/style things separately. Or perhaps, we maintain a HTML file having latest version's updates with proper HTML styles that Adobe AIR HTML component can render. 

When we can use the Adobe AIR's HTML component as it is in the Windows' version, we can't use that API in sandbox macOS version. During application packaging, we need to remove the HTML framework (WebKit.dylib) from packaging else Apple will refuse the application as its said. In that case, if we agreed to go by HTML rendering, we might require to investigate to have `WebView` supported through native extension. 

Please, suggest your thoughts, @piotrzarzycki21 , @JoelProminic .

**Comment by rat-moonshine** [comment]
Having WebView support through ANE in an Adobe AIR project is expensive. Previously, Distriqt tried to do this but their target pledge failed at Kickstarter. 

There are some developers still building features on this using MKWebView on macOS desktop (and other relative platform), one probable destination is this: https://github.com/tuarua/WebViewANE. 

However, when I asked that whether we can build the library with XCode 8.0, they refused as the library has Swift 4.0 usage. Following their one suggestion I tried to download and install XCode 9.3 on El Capitan (as they mentioned one can run multiple version of XCode on a system), but El Capitan refused to run the file saying it needs an upgraded OS.

**Comment by rat-moonshine** [comment]
Here is the response from the developer on asking a compiled source:

> I do provide the ANE compiled.Follow the instructions in the README https://github.com/tuarua/WebViewANE#the-ane--dependencies-1
> 
> It's also in the releases section https://github.com/tuarua/WebViewANE/releases
> You don't need Xcode at all if you just want to use on OSX desktop app.

We can start investigation from this point.

**Comment by rat-moonshine** [comment]
I able to run the supplied example with provided .ane files through Moonshine. 

I able to get a view like this which seems to be a running native web view in macOS, but no website were loading:
![screenshot](https://user-images.githubusercontent.com/24370870/39750348-9e766454-52d2-11e8-9106-f1ff6ec1c569.png)

I reported the issue asking help to the Git https://github.com/tuarua/WebViewANE/issues/147.

**Comment by piotrzarzycki21** [comment]
This issue is related to new one #455 - Probably we will close that one once #455 will be done.
